plugins/vppv2/l2plugin/descriptor/bd_interface.go

```diff
@@ -0,0 +1,135 @@
+
+	// dependency labels
+	interfaceDep = "interface"
+)
+
+// BDInterfaceDescriptor teaches KVScheduler how to put interface into VPP bridge
+// domain.
+type BDInterfaceDescriptor struct {
+	// dependencies
+	log       logging.Logger
+	bdIndex   idxvpp2.NameToIndex
+	bdHandler vppcalls.BridgeDomainVppAPI
+}
+
+// NewBDInterfaceDescriptor creates a new instance of the BDInterface descriptor.
+func NewBDInterfaceDescriptor(bdIndex idxvpp2.NameToIndex, bdHandler vppcalls.BridgeDomainVppAPI, log logging.PluginLogger) *BDInterfaceDescriptor {
+
+	return &BDInterfaceDescriptor{
+		bdIndex:   bdIndex,
+		bdHandler: bdHandler,
+		log:       log.NewLogger("bd-iface-descriptor"),
+	}
+}
+
+// GetDescriptor returns descriptor suitable for registration (via adapter) with
+// the KVScheduler.
+func (d *BDInterfaceDescriptor) GetDescriptor() *adapter.BDInterfaceDescriptor {
+	return &adapter.BDInterfaceDescriptor{
+		Name:               BDInterfaceDescriptorName,
+		KeySelector:        d.IsBDInterfaceKey,
+		ValueTypeName:      proto.MessageName(&l2.BridgeDomain_Interface{}),
+		Add:                d.Add,
+		Delete:             d.Delete,
+		ModifyWithRecreate: d.ModifyWithRecreate,
+		Dependencies:       d.Dependencies,
+	}
+}
+
+// IsBDInterfaceKey returns true if the key is identifying binding between
+// VPP bridge domain and interface.
+func (d *BDInterfaceDescriptor) IsBDInterfaceKey(key string) bool {
+	_, _, isBDIfaceKey := l2.ParseBDInterfaceKey(key)
+	return isBDIfaceKey
+}
+
+// Add puts interface into bridge domain.
+func (d *BDInterfaceDescriptor) Add(key string, bdIface *l2.BridgeDomain_Interface) (metadata interface{}, err error) {
+	// get bridge domain index
+	bdName, _, _ := l2.ParseBDInterfaceKey(key)
+	bdMeta, found := d.bdIndex.LookupByName(bdName)
+	if !found {
+		err = errors.Errorf("failed to obtain metadata for bridge domain %s", bdName)
+		d.log.Error(err)
+		return nil, err
+	}
+
+	// put interface into the bridge domain
+	err = d.bdHandler.AddInterfaceToBridgeDomain(bdMeta.GetIndex(), bdIface)
+	if err != nil {
+		d.log.Error(err)
+		return nil, err
+
+	}
+	return nil, nil
+}
+
+// Delete removes interface from bridge domain.
+func (d *BDInterfaceDescriptor) Delete(key string, bdIface *l2.BridgeDomain_Interface, metadata interface{}) error {
+	// get bridge domain index
+	bdName, _, _ := l2.ParseBDInterfaceKey(key)
+	bdMeta, found := d.bdIndex.LookupByName(bdName)
+	if !found {
+		err := errors.Errorf("failed to obtain metadata for bridge domain %s", bdName)
+		d.log.Error(err)
+		return err
+	}
+
+	err := d.bdHandler.DeleteInterfaceFromBridgeDomain(bdMeta.GetIndex(), bdIface)
+	if err != nil {
+		d.log.Error(err)
+		return err
+
+	}
+	return nil
+}
+
+// ModifyWithRecreate returns always true - a change in BVI or SHG is always performed
+// via Delete+Add.
+func (d *BDInterfaceDescriptor) ModifyWithRecreate(key string, oldBDIface, newBDIface *l2.BridgeDomain_Interface, metadata interface{}) bool {
```

Comment:
Could this be omitted, left nil in descriptor and then scheduler could figure it out automatically when Modify is nil as well.

Reply:
I guess it could, this way it is just explicit that Modify wasn't excluded by mistake,